Add explanation for APF debug output (#42075)
kubernetes/website: content/en/docs/reference/flow-control/_index.md

---
title: Flow control
weight: 120

> no_list: true

Reviewer: Since this is a "no_list" subdirectory, maybe we can create a `reference/debug-cluster` directory and put this page under that directory. When we name this file as `reference/debug-cluster/flow-control.md`, we are making room for future topics related to debugging cluster.

Author: Moved to `reference/debug-cluster`. Please let me know if `weight` and other configuration make sense.

---

<!-- overview -->

API Priority and Fairness controls the behavior of the Kubernetes API server in
an overload situation. You can find more information about it in the
[API Priority and Fairness](/docs/concepts/cluster-administration/flow-control/)
documentation.

<!-- body -->

## Diagnostics

Every HTTP response from an API server with the priority and fairness feature
enabled has two extra headers: `X-Kubernetes-PF-FlowSchema-UID` and
`X-Kubernetes-PF-PriorityLevel-UID`, noting the flow schema that matched the request
and the priority level to which it was assigned, respectively. The API objects'
names are not included in these headers (to avoid revealing details in case the
requesting user does not have permission to view them). When debugging, you
can use a command such as:

```shell
kubectl get flowschemas -o custom-columns="uid:{metadata.uid},name:{metadata.name}"
kubectl get prioritylevelconfigurations -o custom-columns="uid:{metadata.uid},name:{metadata.name}"
```

to get a mapping of UIDs to names for both FlowSchemas and
PriorityLevelConfigurations.

## Debug endpoints

> When you enable the API Priority and Fairness feature, the `kube-apiserver`

Reviewer: ```suggestion
With the `APIPriorityAndFairness` feature enabled, the `kube-apiserver`
```

serves the following additional paths at its HTTP(S) ports.

You need to ensure you have permissions to access these endpoints.
You don't have to do anything if you are using admin.

> Permissions can be granted if needed following [this doc](https://kubernetes.io/docs/reference/access-authn-authz/rbac/)

Reviewer: ```suggestion
Permissions can be granted if needed following the [RBAC](/docs/reference/access-authn-authz/rbac/)
```

  ```

  Explanation for selected column names:
  - `QueueIndex`: The index of the queue. It will be -1 for priority levels
    without queues.
  - `RequestIndexInQueue`: The index in the queue for a given request. It will
    be -1 for executing requests.
  - `InitialSeats`: The number of seats will be occupied during the initial
    (normal) stage of execution of the request.
  - `FinalSeats`: The number of seats will be occupied during the final stage
    of request execution, accounting for the associated WATCH notifications.
  - `AdditionalLatency`: The extra time taken during the final stage of request
    execution. FinalSeats will be occupied during this time period. It does not
    mean any latency that a user will observe.
  - `StartTime`: The time a request starts to execute. It will be
    0001-01-01T00:00:00Z for queued requests.

## Debug logging

> At `-v=3` or more verbosity, the API server outputs an httplog line for every

Reviewer: httplog is a new term to me ...

Author: It is provided by https://github.com/kubernetes/apiserver/blob/master/pkg/server/httplog/doc.go. 
It allows you to add key-value pairs which is heavily used here. https://github.com/kubernetes/apiserver/blob/83bd32f90ba0c7cb082e2605a21e1a9f183481cd/pkg/server/httplog/httplog.go#L227

> request in the APIServer log, and it includes the following attributes.

Reviewer: ```suggestion
request in the API server log, and it includes the following attributes.
```

- `apf_fs`: the name of the flow schema to which the request was classified.
- `apf_pl`: the name of the priority level for that flow schema.
- `apf_iseats`: the number of seats determined for the initial
  (normal) stage of execution of the request.
- `apf_fseats`: the number of seats determined for the final stage of

>   execution (accounting for the associated WATCH notifications) of the

Reviewer: ```suggestion
  execution (accounting for the associated `watch` notifications) of the
```

  request.
- `apf_additionalLatency`: the duration of the final stage of
  execution of the request.

At higher levels of verbosity there will be log lines exposing details
of how APF handled the request, primarily for debugging purposes.

## Response headers

APF adds the following two headers to each HTTP response message.
They won't appear in the audit log. They can be viewed from the client side.

> For client using klog, use verbosity `-v=8` or higher to view these headers.

Reviewer: ```suggestion
For client using `klog`, use verbosity `-v=8` or higher to view these headers.
```
